### ISSUE TITLE
Off-by-one error on frequency information when reading filterbank files

### ISSUE THREAD

**Issue by thevorpalblade** [UCBerkeleySETI/blimpy#118]
**Describe the bug**
When making a Waterfall object from a filterbank file, the .info() command does not correctly report the minimum frequency (but does appear to report the maximum frequency correctly).

**To Reproduce**
1. 
`from blimpy import Waterfall
wf = Waterfall("somefile.fil)
\# I used spliced_blc40414243444546o7o0515253545556o7o0616263646566o7o071727374757677_guppi_58702_18407_BLGCsurvey_Cband_A00_0025.gpuspec.0000.fil
 wf.info()                                                                       

--- File Info ---
      machine_id :                               20
    telescope_id :                                6
       data_type :                                1
            fch1 :            8438.964842353016 MHz
            foff :      -2.7939677238464355e-06 MHz
          nchans :                       1744830464
     source_name :             BLGCsurvey_Cband_A00
         src_raj :                     17:45:39.144
         src_dej :                     -28:59:57.12
        az_start :                              0.0
        za_start :                              0.0
           nbits :                               32
   tstart (ISOT) :          2019-08-07T05:06:47.000
    tstart (MJD) :                58702.21304398148
           tsamp :                     18.253611008
            nifs :                                1

Num ints in file :                               16
      File shape :              (16, 1, 1744830464)
--- Selection Info ---
Data selection shape :              (16, 1, 1744830464)
Minimum freq (MHz) :                3563.964842353016
Maximum freq (MHz) :                8438.964842353016

freqs = wf0.container.populate_freqs()
freqs[-1]
\# returns 3563.964845146984, not  3563.964842353016 as expected
`

**Expected behavior**
I would expect the minimum frequency in the populated array to match that from the file info. I'm not sure which is correct.

**Setup**
Python 3.8.2
Blimpy 2.0.0

**Comment by texadactyl** [comment]
regarding, spliced_blc40414243444546o7o0515253545556o7o0616263646566o7o071727374757677_guppi_58702_18407_BLGCsurvey_Cband_A00_0025.gpuspec.0000.fil
where is it (URL)?

**Comment by thevorpalblade** [comment]
http://blpd9.ssl.berkeley.edu/GC/AGBT19B_999_03/spliced_blc40414243444546o7o0515253545556o7o0616263646566o7o071727374757677_guppi_58702_18407_BLGCsurvey_Cband_A00_0025.gpuspec.0000.fil

Although this one is probably a better choice as it is only 1.7 GB, rather than 100 :-)
I get the same behavior with it (and any of the other randomly selected files I've tried):
http://blpd9.ssl.berkeley.edu/GC/AGBT19B_999_03/spliced_blc40414243444546o7o0515253545556o7o0616263646566o7o071727374757677_guppi_58702_18407_BLGCsurvey_Cband_A00_0025.gpuspec.0002.fil

**Comment by texadactyl** [comment]
I saw a slight discrepancy with http://blpd0.ssl.berkeley.edu/Voyager_data/Voyager1.single_coarse.fine_res.fil
too but the 2 figures (wf.container.f_start from info() and freqs[-1] from populate_freqs()) agree up to 8 digits.  With your test file, they agree up to 9 digits.

What are the accuracy requirements?  @telegraphic ?

Digging into blimpy some more.

**Comment by thevorpalblade** [comment]
To me it looks like likely the result of np.linspace(endpoint=False) instead of np.linspace(endpoint=True) for example, thus my suspicion of an off-by-one. 

To further back this up, the discrepency in both the data files I linked above is exactly the width of a single frequency bin (about 3 Hz in the .0000.fil file and about 3 kHz in the .0002.fil file).

**Comment by texadactyl** [comment]
The discrepancy seems to be caused by the frequency array produced by io/base_reader.populate_freqs():
```
    def populate_freqs(self):
        """
         Populate frequency axis
        """

        if self.header['foff'] < 0:
            f0 = self.f_end
        else:
            f0 = self.f_begin

        self._setup_chans()

        #create freq array
        i_vals = np.arange(self.chan_start_idx, self.chan_stop_idx)
        freqs = self.header['foff'] * i_vals + f0

        return freqs

```

Change 
```
i_vals = np.arange(self.chan_start_idx, self.chan_stop_idx)
```
to
```
i_vals = np.arange(self.chan_start_idx, self.chan_stop_idx + 1)
```
and there is no more discrepancy with the Voyager 1 FIL file.

Max always matched because it was located at index=0 in the array.

**Comment by texadactyl** [comment]
@thevorpalblade  Please verify if you can.  I'll create a pull request if you do.

**Comment by texadactyl** [comment]
@thevorpalblade  Alternatively to you modifying the code, you are welcome to download my branch patch-1 of blimpy:  https://github.com/texadactyl/blimpy/tree/patch-1

**Comment by texadactyl** [comment]
Found a good source of FIL test data: http://seti.berkeley.edu/opendata
and I was able to test several FIL files, all successful.

**Comment by texadactyl** [comment]
Apologies.  The clone texadactyl/blimpy may have taken an incorrect approach.  Will followup.

**Comment by telegraphic** [comment]
Thanks @thevorpalblade -- not the first off-by-X issue we've had with spectra!  Will have to find some time to look into it properly, but agree that it looks like a bug, so labelling as such

**Comment by thevorpalblade** [comment]
Do you have a sense of which value is the correct one for the minimum frequency? For now I can just create my own frequency axis as a workaround.

**Comment by texadactyl** [comment]
@thevorpalblade 
This issue has been addressed.
Please give this a test again when you are able.
I believe that you will now see that freqs[-1] = minimum frequency.
If you are in agreement, please close this issue.  Thnaks.

**Comment by thevorpalblade** [comment]
Yep, this works! Thanks!

**Comment by thevorpalblade** [comment]
So, I just noticed that now the result of `populate_freqs()` is one longer than the number of points in the spectrum, whereas I do not think that this was the case before. If the frequencies are the edges of the frequency bins this makes sense...but since this is an FFT I would expect that it's more the centers of the frequency bins, which means (I think) that the patch introduced a regression. Should I reopen the issue? Post a new one? Or is it not actually an issue?

**Comment by telegraphic** [comment]
Calling @david-macmahon for FFT input!

**Comment by david-macmahon** [comment]
It looks like this was not a bug to begin with.  `fch1` is the center frequency of the first channel in the filterbank file.  `foff` is the frequency step (offset) from the center of one channel to the center of the channel after it.  `foff` will be negative if the channels are ordered from high to low, as in the example shown in the initial report. The frequency of the last channel in the filterbank file is calculated as `fch1 + foff*(nchans-1)`.  For the example shown above, we have:
```
In [1]: fch1=8438.964842353016
In [2]: foff=-2.7939677238464355e-06
In [3]: nchans=1744830464
In [4]: fch1 + foff*(nchans-1)
Out[4]: 3563.964845146984
```
which is the last value of the array returned by the original `populate_freqs()`.

As a simplified case, imagine `fch1=0`, `foff=1`, `nchans=2`.  The returned frequency array should be `[0, 1]`, **not** `[0, 2]` or `[0, 1, 2]`.

I think this "fix" should be reverted.

**Comment by texadactyl** [comment]
I created a fork at https://github.com/texadactyl/blimpy/tree/patch-1
@telegraphic:
1) I reversed the "+ 1"s in the index values for populate_freqs(), in grab_data() where it is calculating plot_f and plot_data, and in tests/test_voyager_data_load.py.
2) I left the other 2 fixes in setup.py and calib_plots.py.
3) pytest passes and my own unit tests work fine too on both .fil and h5 of voyager.

But, we are back to that point where wf.container.f_start as reported by for info() is not the minimum of the frequency array returned by populate_freqs().

info():  Minimum freq (MHz): 8418.457029853016  <--- wf.container.f_start
np.min(populate_freqs()):     8418.457032646984

(cont.)

**Comment by texadactyl** [comment]
My test program:
```
import numpy as np
from blimpy import Waterfall

FILFILE = "Voyager1.single_coarse.fine_res.fil"
DATADIR = "/mnt/elkdata/seti/data/voyager/"
BLIMPY_FILE = DATADIR + FILFILE

wf = Waterfall(BLIMPY_FILE)
print("wf.container:\n", vars(wf.container))

freqs = wf.container.populate_freqs()
print("\nfreqs({}):\n{}".format(len(freqs), freqs))
print(wf.info())
minfreq = np.min(freqs)
maxfreq = np.max(freqs)
print("freq array:\nfirst={}, last={}, min={}, max={}, f_start={}, f_stop={}"
      .format(freqs[0], freqs[-1], minfreq, maxfreq, wf.container.f_start,
              wf.container.f_stop))

ix = 0
for freq in freqs:
    if freq == minfreq:
        print("Found minfreq at index ", ix)
    if freq == maxfreq:
        print("Found maxfreq at index ", ix)
    if freq == wf.container.f_start:
        print("Found wf.container.f_start at index ", ix)
    if freq == wf.container.f_stop:
        print("Found wf.container.f_stop at index ", ix)
    if freq == wf.container.f_begin:
        print("Found wf.container.f_begin at index ", ix)
    if freq == wf.container.f_end:
        print("Found wf.container.f_end at index ", ix)
    ix += 1

if minfreq == wf.container.f_start:
    print("minfreq equals wf.container.f_start")
if maxfreq == wf.container.f_stop:
    print("maxfreq equals wf.container.f_stop")

```

**Comment by texadactyl** [comment]
Corrected info().  
```
        if self.header['foff'] < 0:
            minfreq = self.container.f_start - self.header['foff']
            maxfreq = self.container.f_stop
        else:
            minfreq = self.container.f_start
            maxfreq = self.container.f_stop - self.header['foff']
        print("%16s : %32s" % ("Minimum freq (MHz)", minfreq))
        print("%16s : %32s" % ("Maximum freq (MHz)", maxfreq))

```
**Generated two simpler files than voyager.**

nchans = 5, fch1 = 100, foff = -2
freqs = [100, 98, 96, 94, 92]
Waterfall:
a) f_start = 90, f_stop = 100
b) Corrected info() shows the min=92 and max=100
c) f_start is not part of frequency array.

nchans = 5, fch1 = 100, foff = +2
freqs = [100, 102, 104, 106, 108]
Waterfall:
a) f_start = 100, f_stop = 110
b) Corrected info() shows the right min=100 and max=108.
c) f_stop is not part of frequency array.

**Voyager**

nchans = 1048576, fch1 = 8421.386717353016, foff = -2.7939677238464355e-06
freqs(1048576):
[8421.38671735 8421.38671456 8421.38671177 ... 8418.45703823 8418.45703544
 8418.45703265]

--- Selection Info ---
Data selection shape :                 (16, 1, 1048576)
Minimum freq (MHz) :                8418.457032646984
Maximum freq (MHz) :                8421.386717353016

freq array:
first=8421.386717353016, last=8418.457032646984, min=8418.457032646984, max=8421.386717353016, f_start=8418.457029853016, f_stop=8421.386717353016

**Comment by david-macmahon** [comment]
Getting closer.  The array of center frequencies is now correct, but I think the start/stop frequencies are off by half a channel width (and possibly swapped, depending on definitions).

> nchans = 5, fch1 = 100, foff = -2
freqs = [100, 98, 96, 94, 92]
Waterfall:
a) f_start = 90, f_stop = 100
b) Corrected info() shows the min=92 and max=100
c) f_start is not part of frequency array.

The first channel is centered at 100 and has a bandwidth of -2, so that channel spans frequency range 101 to 99.  Likewise, the last channel spans from 93 to 91.

I don't know the actual definition of `f_start` and `f_stop`, but assuming they are the edges of the band, they should be either 101 or 91.  I'm not sure whether `f_start` (`f_stop`) is the "outer" edge of the _first_ (last) channel or the "outer" edge of the _lowest_ (highest) frequency channel.

**Comment by texadactyl** [comment]
Thanks, @david-macmahon 

As far as I know, they're all center points in the list of frequencies.  Keep in mind that my knowledge of blimpy is based on reading the source code.  @telegraphic can speak authoritatively about blimpy requirements with respect to telescope data from guppi files.

My comments of variables f_start etc. probably have more meaning internally to blimpy.  The frequency arrays that I depicted show the actual order of the frequencies.

**Comment by telegraphic** [comment]
@david-macmahon so I think `f_start` *should* refer to the central frequency of the first channel.  The DC bin (zeroth channel) output from the FFT is centred on 0 +/- 1/2 bin width -- with the negative frequencies aliasing to the highest frequency.

If agreed, documentation should spell this out specifically, and grab_data functions with `f_start` argument should refer to bin centers.

**Comment by david-macmahon** [comment]
The `f_start` and `f_stop` parameters can be given as command line arguments to `watutil` (and others?).  I could be wrong, but they don't ever seem to be adjusted to ensure that they are a suitably distanced from from `fch1` so it's hard to tell what the intent is.  It looks like channel indices get calculated by rounding `(f_start - fch1) / foff` to an integer value, so presumably all `f_start` values that are within half a channel width of the channel's center frequency will select the same channel.  I haven't checked what value gets written out for `fch1` when slicing by `f_start` and `f_stop`.

**Comment by texadactyl** [comment]
Somewhere along the way, this was fixed as observed in blimpy 2.0.32.
```
from blimpy import Waterfall
from tests.data import voyager_fil

wf = Waterfall(voyager_fil) # from http://blpd0.ssl.berkeley.edu/Voyager_data/Voyager1.single_coarse.fine_res.fil

wf.info()
freqs = wf.container.populate_freqs()

print("\nLast one in the frequency array:", freqs[-1])

len_f = len(freqs)
first = wf.header["fch1"]
last_1 = freqs[-1]
last_2 = first + (len_f - 1 ) *  wf.header["foff"]

print("Last one (calculated):", last_2)

print("Diff:", abs(last_2 - last_1))
```
produces this output:
```
--- File Info ---
    telescope_id :                                6
        az_start :                              0.0
      machine_id :                               20
     source_name :                         Voyager1
       data_type :                                1
          nchans :                          1048576
           ibeam :                                1
           tsamp :                     18.253611008
            foff :      -2.7939677238464355e-06 MHz
         src_raj :                     17:10:03.984
         src_dej :                       12:10:58.8
   tstart (ISOT) :          2016-09-19T18:46:13.000
    tstart (MJD) :                57650.78209490741
          nbeams :                                1
            fch1 :            8421.386717353016 MHz
        za_start :                              0.0
     rawdatafile : guppi_57650_67573_Voyager1_0002.0000.raw
            nifs :                                1
           nbits :                               32

Num ints in file :                               16
      File shape :                 (16, 1, 1048576)
--- Selection Info ---
Data selection shape :                 (16, 1, 1048576)
Minimum freq (MHz) :                8418.457032646984
Maximum freq (MHz) :                8421.386717353016

Last one in the frequency array: 8418.457032646984
Last one (calculated): 8418.457032646984
Diff: 0.0
```
